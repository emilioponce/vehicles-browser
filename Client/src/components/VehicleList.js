import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Vehicle from './Vehicle'
import Pagination from 'react-js-pagination'
import styles from './VehicleList.module.css'
import AppContext from '../config/context'
import { PAGE_RANGE, ITEMS_PER_PAGE } from '../config/constants'

class VehicleList extends Component {
  constructor(props) {
    super(props)
    this.state = {
      activePage: 1,
      pageItems: []
    }
  }

  getOffset = page => (page - 1) * ITEMS_PER_PAGE

  handlePageChange = page => {
    this.setState({ activePage: page })
    let offset = this.getOffset(page)
    this.setState({
      pageItems: this.props.vehicles.slice(offset, offset + ITEMS_PER_PAGE)
    })
  }

  componentDidUpdate(prevProps, prevState) {
    //Updating pageItems in context Provider, as the map will consum this info
    if (this.state.pageItems !== prevState.pageItems) {
      this.props.updateVisibleCars(this.state.pageItems)
    }
    // if props are update, we update also the state
    if (this.props.vehicles !== prevProps.vehicles) {
      this.setState({ pageItems: this.props.vehicles.slice(0, ITEMS_PER_PAGE) })
    }
  }

  render() {
    const { activePage, pageItems } = this.state
    const { vehicles } = this.props

    if (pageItems.length === 0) {
      return <div className={styles.vehicleList}>Loading...</div>
    }

    return (
      <div className={styles.vehicleList}>
        <h3>Vehicle List</h3>
        <div>
          {pageItems.map(vehicle => (
            <Vehicle key={vehicle.id} {...vehicle} />
          ))}
        </div>

        <div className={styles.paginationWrapper}>
          <Pagination
            activePage={activePage}
            itemsCountPerPage={ITEMS_PER_PAGE}
            totalItemsCount={vehicles.length}
            pageRangeDisplayed={PAGE_RANGE}
            onChange={this.handlePageChange}
          />
        </div>
      </div>
    )
  }
}

VehicleList.propTypes = {
  vehicles: PropTypes.array.isRequired,
  updateVisibleCars: PropTypes.func
}

// To pass prop updateVisibleCars to VehicleList before component initialization
const ConnectedVehicleListUpdate = props => (
  <AppContext.Consumer>
    {({ updateVisibleCars }) => (
      <VehicleList {...props} updateVisibleCars={updateVisibleCars} />
    )}
  </AppContext.Consumer>
)

export default ConnectedVehicleListUpdate
