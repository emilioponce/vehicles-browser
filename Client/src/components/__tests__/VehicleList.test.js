import React from 'react'
import ReactDOM from 'react-dom'
import VehicleList from '../VehicleList'
import renderer from 'react-test-renderer'

const component = <VehicleList vehicles={[]} updateVisibleCars={jest.fn()} />

describe('Component VehicleList', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(component, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  it('snapshot renders correctly', () => {
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
