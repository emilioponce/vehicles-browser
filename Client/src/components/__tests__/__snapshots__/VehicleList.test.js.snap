// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component VehicleList snapshot renders correctly 1`] = `
<div
  className="vehicleList"
>
  Loading...
</div>
`;
